pep-0695.rst

```diff
@@ -302,6 +302,20 @@ the existing rules enforced by type checkers for a ``TypeVar`` constructor call.
     class ClassG[T: (list[S], str)]: ...  # Type checker error: generic type
 
 
+Runtime Representation of Bounds and Constraints
+------------------------------------------------
+
+The upper bounds and constraints of ``TypeVar`` objects are accessible at
+runtime through the ``__bound__`` and ``__constraints__`` attributes.
+For ``TypeVar`` objects defined through the new syntax, these attributes
+become lazily evaluated: when constructed, the object stores a closure that
+allows the evaluation of the value. When the ``__bound__`` or ``__constraints__``
+attribute is accessed, the value is evaluated and cached. This is consistent
+with the behavior of :pep:`649` and generally eliminates the need for quoted
+forward references. If :pep:`649` is accepted, the ``TypeVar`` class should
```

Comment:
I don't understand that last sentence. Are you saying that the `TypeVar` class won't support deferred evaluation if PEP 649 is not accepted?

Reply:
It may be worth clarifying that deferred evaluation will change the behavior in cases where the symbol is assigned different values. For example:

```python
X = int

def class Foo[T: X]:
    # print(T.__bound__)

X = str
```

With immediate evaluation, the bound will be `int`. With deferred evaluation, it will be `str` unless the `print` statement is uncommented, in which case it will still be `int`. I realize this is an edge case. I also realize it affects all deferred evaluation approaches, including deferred annotations using today's `from __future__ import annotations` mechanism.

Mental note to me: Pyright's current implementation of PEP 695 reports an error if the bound or constraints refers to a symbol that is not yet defined but is not quoted. I'll need to change this.

Reply:
My intent here is that if/when PEP 649 is implemented, we add something like an `__evaluate_bound__(self, format)` method, similar to the `__annotate__` method in PEP 649 (https://peps.python.org/pep-0649/#annotate-and-annotations). This will allow evaluating the bound to the "STRING" or "FORWARDREF" format. I will try to clarify that.

And yes, deferred evaluation can affect the behavior. I think I should add a section on "Deferred evaluation" to cover a number of these subtleties.

Reply:
Added an extended discussion of lazy evaluation.